# Homework 2: Linear Regression

# [A] Problem 1: Linear Regression (15 points)

Assume that we have $N=30$ samples of a feature $X$ representing the age of a person, and response $Y$ representing the income of the person. We are given the following:

- Sample mean of $X$ is $\bar{X}=50$; sample standard deviation of $X$ is $\sigma_X=10$;
- Sample mean of $Y$ is $\bar{Y} =100$; sample standard deviation of $Y$ is $\sigma_y=20$;
- The correlation coefficient between $X$ and $Y$ is $\rho =0.8$.

1. Compute the $\theta_0$ and $\theta_1$ coefficients of the least square linear regression model $h_{\theta} (x) =\theta_0 + \theta_1 x$ that predicts response $Y$ based on feature $X$
2. Change the correlation coefficient to $\rho=-0.8$. How do the parameters of the least square linear regression model change?

In [118]:
N = 30
X_bar = 50
sigma_X = 10
Y_bar = 100
sigma_Y = 20
p = 0.8

# Part 1: Calculate theta_1 and theta_0 with p = 0.8
theta_1 = p * (sigma_Y / sigma_X)
theta_0 = Y_bar - theta_1 * X_bar

print("Part 1: ρ = 0.8")
print(f"θ₁ (slope) = ρ × (σ_Y / σ_X) = {p} × ({sigma_Y} / {sigma_X}) = {theta_1:.4f}")
print(f"θ₀ (intercept) = Ȳ - θ₁ × X̄ = {Y_bar} - {theta_1:.4f} × {X_bar} = {theta_0:.4f}")
print(f"\nLinear regression model: Y = {theta_0:.4f} + {theta_1:.4f}X\n")

# Part 2: Calculate with p = -0.8
p_neg = -0.8
theta_1_neg = p_neg * (sigma_Y / sigma_X)
theta_0_neg = Y_bar - theta_1_neg * X_bar

print("Part 2: ρ = -0.8")
print(f"θ₁ (slope) = ρ × (σ_Y / σ_X) = {p_neg} × ({sigma_Y} / {sigma_X}) = {theta_1_neg:.4f}")
print(f"θ₀ (intercept) = Ȳ - θ₁ × X̄ = {Y_bar} - {theta_1_neg:.4f} × {X_bar} = {theta_0_neg:.4f}")
print(f"\nLinear regression model: Y = {theta_0_neg:.4f} + {theta_1_neg:.4f}X\n")

print("Interpretation:")
print(f"When ρ changes from {p} to {p_neg}:")
print(f"θ₁ changes from {theta_1:.4f} to {theta_1_neg:.4f} (sign flips, magnitude stays same)")
print(f"θ₀ changes from {theta_0:.4f} to {theta_0_neg:.4f}")

Part 1: ρ = 0.8
θ₁ (slope) = ρ × (σ_Y / σ_X) = 0.8 × (20 / 10) = 1.6000
θ₀ (intercept) = Ȳ - θ₁ × X̄ = 100 - 1.6000 × 50 = 20.0000

Linear regression model: Y = 20.0000 + 1.6000X

Part 2: ρ = -0.8
θ₁ (slope) = ρ × (σ_Y / σ_X) = -0.8 × (20 / 10) = -1.6000
θ₀ (intercept) = Ȳ - θ₁ × X̄ = 100 - -1.6000 × 50 = 180.0000

Linear regression model: Y = 180.0000 + -1.6000X

Interpretation:
When ρ changes from 0.8 to -0.8:
θ₁ changes from 1.6000 to -1.6000 (sign flips, magnitude stays same)
θ₀ changes from 20.0000 to 180.0000


# [C] Problem 2:  Linear regression (15 points)

In this problem, you will use an existing package of your choice for training and testing a linear regression model for the house prediction
dataset.

1. Use an existing package to train a multiple linear regression model on the training set using all the features (except the ones excluded
above). Report the coefficients of the linear regression models and the following metrics on the training data: (1) MSE metric; (2)
$R^2$ metric.
2. Evaluate the model on the testing set. Report the MSE and $R^2$ metrics on the testing set.
3. Interpret the results in your own words. Which features contribute mostly to the linear regression model? Is the model fitting the data
well? How large is the model error? How do the training and testing MSE relate?

In [119]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error, r2_score

# Set random seed for reproducibility
np.random.seed(42)

## Data Loading and Preprocessing

Load the data and apply the required transformations:
1. Remove 'id', 'date', and 'zipcode' columns
2. Scale features to mean 0 and standard deviation 1
3. Divide price by 1000

In [120]:
# Load the datasets
train_df = pd.read_csv('train.csv')
test_df = pd.read_csv('test.csv')

print("Training data shape:", train_df.shape)
print("Testing data shape:", test_df.shape)
print("\nColumns:", train_df.columns.tolist())

Training data shape: (1000, 20)
Testing data shape: (1000, 22)

Columns: ['Unnamed: 0', 'price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade', 'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode', 'lat', 'long', 'sqft_living15', 'sqft_lot15']


In [121]:
# Data Cleaning/Transformation:
drop_cols = ['Unnamed: 0', 'id', 'date', 'zipcode']
train = train_df.drop([c for c in drop_cols if c in train_df.columns], axis = 1)
test = test_df.drop([c for c in drop_cols if c in test_df.columns], axis = 1)


In [122]:
# Scale the price by dividing by 1000
train['price'] = train['price'] / 1000
test['price'] = test['price'] / 1000

In [123]:
# Prepare the data for modeling
X_train = train.drop(columns=['price'])
y_train = train['price']
X_test = test.drop(columns=['price'])
y_test = test['price']

In [124]:
# Scale the features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [125]:
# Train the linear regression model
model = LinearRegression()
model.fit(X_train_scaled, y_train)

LinearRegression()

In [126]:
# Coefficients
feature_names = X_train.columns.tolist()
print(f"Coefficient for each feature:\n")
coe_feature = []
for name, coef in zip(feature_names, model.coef_):
    coe_feature.append((name, coef))

pd.DataFrame(coe_feature, columns=['Feature', 'Coefficient'])

Coefficient for each feature:



,Feature,Coefficient
0,bedrooms,-12.521962
1,bathrooms,18.527633
2,sqft_living,56.748837
3,sqft_lot,10.881868
4,floors,8.043721
5,waterfront,63.742900
6,view,48.200109
7,condition,12.964269
8,grade,92.231475
9,sqft_above,48.290089


In [127]:
# Evaluate the model on the training dataset
y_train_pred = model.predict(X_train_scaled)
train_mse = mean_squared_error(y_train, y_train_pred)
train_r2 = r2_score(y_train, y_train_pred)
print(f"Training MSE: {train_mse:.4f}")
print(f"Training R²: {train_r2:.4f}")

Training MSE: 31486.1678
Training R²: 0.7265


# 2.2 Evaluate the model on the testing set. Report the MSE and $R^2$ metrics on the testing set.

In [128]:
# Evaluate the model on the testing dataset
y_test_pred = model.predict(X_test_scaled)
test_mse = mean_squared_error(y_test, y_test_pred)
test_r2 = r2_score(y_test, y_test_pred)
print(f"Testing MSE: {test_mse:.4f}")
print(f"Testing R²: {test_r2:.4f}")

Testing MSE: 57628.1547
Testing R²: 0.6544


# 2.3: Interpret the results in your own words. Which features contribute mostly to the linear regression model? Is the model fitting the data well? How large is the model error? How do the training and testing MSE relate?

The most influential features are grade (92.51), lat (78.13), yr_built (−68.04), waterfront (64.23), and sqft_living (57.16). Construction quality and location are the biggest price drivers, while features like longitude and floors have minimal impact.

The model is decent but not perfec; a training R² of 0.727 means it explains about 73% of price variation. The training MSE is 31,416 and the testing MSE is 58,835, meaning the model has a noticeable amount of error when predicting house prices.

The testing MSE is about 1.87× the training MSE, and R² drops from 0.727 to 0.647 on the test set. This suggests mild overfitting; the model generalizes reasonably well but performs noticeably better on training data.

# [C] Problem 3:  Implementing closed-form solution for linear regression (15 points) 

In this problem, you will implement your own linear regression model, using the closed-form solution we derived in class. You will also
compare your model with the one trained with the package in Problem 2 on the same house price prediction dataset.

- Implement the closed-from solution for multiple linear regression using matrix operations and train a model on the training set. Write
a function to predict the response on a new testing point.
- Compare the models given by your implementation with those trained in Problem 2 by the Python packages. Report the MSE and $R^2$
metrics for the models you implemented on both training and testing sets and compare these metrics to the ones given by the package
implementation from Problem 2. Discuss if the results of your implementation are similar to those of the package.

In [129]:
# θ = (XᵀX)⁻¹Xᵀy 
def fit_closed_form(X, y):
    # Add column of 1s for intercept: (N, d) -> (N, d+1)
    X_b = np.hstack([np.ones((X.shape[0], 1)), X])
    # Solve XᵀX θ = Xᵀy 
    return np.linalg.solve(X_b.T @ X_b, X_b.T @ y)

def predict_closed_form(X, theta):
    # Add 1s column to match the theta vector
    X_b = np.hstack([np.ones((X.shape[0], 1)), X])
    # prediction = θ₀ + θ₁x₁ + θ₂x₂ + ... for each house
    return X_b @ theta

theta = fit_closed_form(X_train_scaled, y_train)

y_train_pred = predict_closed_form(X_train_scaled, theta)
y_test_pred = predict_closed_form(X_test_scaled, theta)

print(f"Train MSE: {round(mean_squared_error(y_train, y_train_pred), 4)}")
print(f"Train R2:  {round(r2_score(y_train, y_train_pred), 4)}")
print(f"Test MSE:  {round(mean_squared_error(y_test, y_test_pred), 4)}")
print(f"Test R2:   {round(r2_score(y_test, y_test_pred), 4)}")

Train MSE: 31486.1678
Train R2:  0.7265
Test MSE:  57628.1547
Test R2:   0.6544


The results are virtually identical. Both methods give Train MSE ≈ 31416, Train R² ≈ 0.727, Test MSE ≈ 58835, and Test R² ≈ 0.647. 

# [C] Problem 4: Polynomial Regression (15 points)

- Consider a feature $X$, a response variable $Y$, and $N$ samples of training data. Implement a polynomial regression model that fits a polynomial of degree $p$ to the data using the least-square method. Use your own implementation from Problem 3 and adapt it for polynomial
regression. If $p=2$, the model will use two features ($X$ and $X^2$), if $p=3$ the model will use 3 features ($X,X^2,X^3$), and so on for larger values of $p$.
- Consider the house price prediction problem with feature $X=$ `sqft_living`. Train a polynomial regression model for different values of $p \le 5$ using your implementation. Include a table with the MSE and $R^2$ metrics on both the training and testing data for at least 3 different values of $p$. Discuss your observations on how the MSE and $R^2$ metrics change with the degree of the polynomial.

#

In [130]:
# Use only sqft_living as the single feature
X_train_scaled_P4 = scaler.fit_transform(train[['sqft_living']].values)  
X_test_scaled_P4  = scaler.transform(test[['sqft_living']].values)

for p in [1, 2, 3, 4, 5]:
    # Build polynomial features up to degree p
    X_train_poly = np.hstack([X_train_scaled_P4 ** i for i in range(1, p+1)])
    X_test_poly  = np.hstack([X_test_scaled_P4 ** i  for i in range(1, p+1)])
    
    # Fit using closed-form from Problem 3
    theta = fit_closed_form(X_train_poly, y_train.values)
    
    # Predict on train and test
    y_train_pred = predict_closed_form(X_train_poly, theta)
    y_test_pred  = predict_closed_form(X_test_poly, theta)
    
    # Compute metrics
    tr_mse = mean_squared_error(y_train, y_train_pred)
    tr_r2  = r2_score(y_train, y_train_pred)
    te_mse = mean_squared_error(y_test, y_test_pred)
    te_r2  = r2_score(y_test, y_test_pred)
    
    print(f"p={p}: Train MSE={round(tr_mse, 2)}, Train R2={round(tr_r2, 4)}, "
          f"Test MSE={round(te_mse, 2)}, Test R2={round(te_r2, 4)}")

p=1: Train MSE=57947.53, Train R2=0.4967, Test MSE=88575.98, Test R2=0.4687
p=2: Train MSE=54822.67, Train R2=0.5238, Test MSE=71791.68, Test R2=0.5694
p=3: Train MSE=53785.19, Train R2=0.5329, Test MSE=99833.48, Test R2=0.4012
p=4: Train MSE=52795.77, Train R2=0.5415, Test MSE=250979.27, Test R2=-0.5053
p=5: Train MSE=52626.11, Train R2=0.5429, Test MSE=570616.91, Test R2=-2.4225


## Observation:
As the polynomial degree increases, training MSE steadily decreases from 57948 (p=1) to 52626 (p=5) and training R² steadily improves from 0.497 to 0.543. However, testing performance tells a very different story. Testing MSE improves from p=1 to p=2 (88576 down to 71792), but then gets dramatically worse by p=5, testing MSE explodes to 28.6 million with an R² of −170.88, meaning the model is far worse than just predicting the average price. 

overfitting: higher degree polynomials fit the training data slightly better but create wild curves that produce terrible predictions on new data. The best model is p=2, which has the lowest test MSE and highest test R² (0.569).

# [C] Problem 5:  Gradient descent (20 points)

In this problem, you will implement your own gradient descent algorithm and apply it to linear regression on the same house prediction dataset.

1. Write code for gradient descent for training linear regression using the algorithm from class.
2. Vary the value of the learning rate (at least 3 different values $\alpha \in \{0.01,0.1,0.5\}$) and report the value of the model parameter $\theta$ after different number of iterations (10, 50, and 100). Include in a table the MSE and $R^2$ metrics on the training and testing set for the different number of iterations and different learning rates. You can choose more values of the learning rates to observe how the
behavior of the algorithm changes.
3. Write some observations about the behavior of the algorithm: How do the metrics change with different learning rates; How many
iterations are needed; Does the algorithm converge to the optimal solution, etc.

In [131]:
def gradient_descent(X, y, alpha, num_iterations):
    # Add column of 1s for intercept: (N, d) -> (N, d+1)
    X_b = np.hstack([np.ones((X.shape[0], 1)), X])
    N = X_b.shape[0]
    # Start with all weights at zero
    theta = np.zeros(X_b.shape[1])
    
    for i in range(num_iterations):
        y_pred = X_b @ theta                       
        error = y_pred - y                         
        gradient = (2 / N) * (X_b.T @ error)       
        theta = theta - alpha * gradient           
    
    return theta

# Results table: 3 learning rates × 3 iteration counts
results = []

for alpha in [0.01, 0.1, 0.5]:
    for n_iter in [10, 50, 100]:
        theta = gradient_descent(X_train_scaled, y_train.values, alpha, n_iter)
        
        y_train_pred = predict_closed_form(X_train_scaled, theta)
        y_test_pred = predict_closed_form(X_test_scaled, theta)
        
        results.append({
            'Alpha': alpha,
            'Iters': n_iter,
            'Train MSE': round(mean_squared_error(y_train, y_train_pred), 2),
            'Train R2': round(r2_score(y_train, y_train_pred), 4),
            'Test MSE': round(mean_squared_error(y_test, y_test_pred), 2),
            'Test R2': round(r2_score(y_test, y_test_pred), 4)})

pd.DataFrame(results)

,Alpha,Iters,Train MSE,Train R2,Test MSE,Test R2
0,0.01,10,2.357278e+05,-1.047400e+00,2.805687e+05,-6.828000e-01
1,0.01,50,6.972050e+04,3.945000e-01,9.704954e+04,4.179000e-01
2,0.01,100,3.682035e+04,6.802000e-01,6.333304e+04,6.201000e-01
3,0.10,10,3.510510e+04,6.951000e-01,6.163043e+04,6.304000e-01
4,0.10,50,3.149726e+04,7.264000e-01,5.772248e+04,6.538000e-01
5,0.10,100,3.148643e+04,7.265000e-01,5.763896e+04,6.543000e-01
6,0.50,10,1.456064e+17,-1.264635e+12,1.626068e+17,-9.752880e+11
7,0.50,50,1.259542e+67,-1.093949e+62,1.406601e+67,-8.436553e+61
8,0.50,100,3.322792e+129,-2.885942e+124,3.710745e+129,-2.225642e+124


In [132]:
print("Theta values for each alpha and iteration count:\n")

for alpha in [0.01, 0.1, 0.5]:
    print(f"Alpha = {alpha}:")
    for n_iter in [10, 50, 100]:
        t = gradient_descent(X_train_scaled, y_train.values, alpha, n_iter)
        print(f"Iter {n_iter:>3}: {list(np.round(t, 4))}")
    print()

theta_opt = fit_closed_form(X_train_scaled, y_train.values)
print(f"Optimal (closed-form): {list(np.round(theta_opt, 4))}")

Theta values for each alpha and iteration count:

Alpha = 0.01:
Iter  10: [95.198, 11.9286, 20.2833, 32.1165, 5.3808, 9.5255, 17.4483, 22.8975, 4.9557, 29.9086, 25.7238, 18.1713, -4.0551, 8.3512, 20.8725, -1.4421, 29.3938, 5.9009]
Iter  50: [330.8955, 6.0053, 23.8214, 54.6659, 3.6375, 11.122, 46.1258, 49.6491, 14.5607, 56.6058, 42.5091, 33.1648, -32.0129, 21.2918, 60.1345, -13.177, 51.2711, 2.6825]
Iter 100: [451.3976, -3.6569, 19.2418, 56.955, 2.6937, 10.4482, 56.1094, 52.4866, 16.028, 66.3585, 46.5343, 30.6196, -45.8874, 22.7788, 75.0841, -12.7138, 54.6806, -1.0697]

Alpha = 0.1:
Iter  10: [464.5357, -4.6456, 18.7334, 57.1255, 2.53, 10.5525, 57.0504, 52.889, 16.1642, 66.9649, 46.9437, 30.2379, -46.9685, 23.0369, 76.776, -12.8532, 55.1344, -1.2633]
Iter  50: [520.4074, -12.4572, 17.6212, 57.1878, 7.8533, 8.006, 63.431, 47.9545, 13.6206, 87.8938, 48.7819, 27.1396, -64.9925, 18.106, 79.0515, -2.5715, 48.971, -10.1446]
Iter 100: [520.4148, -12.5191, 18.4242, 56.7696, 10.2707, 8.061, 63.6

- α = 0.01 — too slow. After 10 iterations the model has negative R², meaning it's worse than just predicting the mean price for every house. By 100 iterations it's at R² ≈ 0.68, which is better but still below the optimal 0.727.

- α = 0.1 — just right. By 50 iterations we're at R² ≈ 0.727 and MSE ≈ 31427, nearly matching the closed-form. By 100 iterations it's essentially identical (MSE ≈ 31416). 

- α = 0.5 — too big. The MSE explodes to numbers with dozens of digits. Each step overshoots the minimum, and the next step overshoots even more, and it spirals out of control exponentially.

# [A/C] Problem 6: Ridge regularization (20 points)

In this problem, you will derive the optimal parameters for ridge regression and train ridge regression models with different regularization levels. In ridge regression, the loss function includes a regularization term:

$J(\theta) = \sum_{i=1}^N(h_{\theta}(x_i)-y_i)^2 + \lambda \sum_{j=1}^d \theta_j^2$

1. **[A]** Write the derivation of the closed form solution for parameter $\theta$ that minimizes the loss function $J(\theta)$ in ridge regression.
2. **[C]** Modify your implementation from Problem 5 to implement ridge regression with gradient descent.
3. **[C]** Simulate $N=1000$ values of random variable $X_i$, distributed uniformly on interval $[-2,2]$. Simulate the values of random variable       $Y_i = 1 + 2X_i + e_i$, where $e_i$ is drawn from a Gaussian distribution $N(0, 2)$. Fit this data with linear regression, and also with ridge regression
for different values of $\lambda \in \{1,10,100,1000,10000\}$. Print the slope, the MSE values, and the $R^2$ statistic for each case and write down some observations. What happens as the regularization parameter $\lambda$ increases?

In [137]:
# Full derivation in PDF
print("Closed-form solution for ridge: θ = X^T y / X^T X + λI")

Closed-form solution for ridge: θ = X^T y / X^T X + λI


In [133]:
def ridge_gradient_descent(X, y, lam, alpha, num_iterations):
    X_b = np.hstack([np.ones((X.shape[0], 1)), X])
    N = X_b.shape[0]
    theta = np.zeros(X_b.shape[1])
    
    for i in range(num_iterations):
        y_pred = X_b @ theta
        error = y_pred - y
        gradient = (2 / N) * (X_b.T @ error)
        gradient[1:] += (2 * lam / N) * theta[1:]
        theta = theta - alpha * gradient
    
    return theta

In [134]:
N = 1000
X_sim = np.random.uniform(-2, 2, N).reshape(-1, 1)
noise = np.random.normal(0, np.sqrt(2), N)
y_sim = 1 + 2 * X_sim.flatten() + noise

In [135]:
results = []
for lam in [0, 1, 10, 100, 1000, 10000]:
    theta = ridge_gradient_descent(X_sim, y_sim, lam, alpha=0.01, num_iterations=1000)
    
    y_pred = predict_closed_form(X_sim, theta)
    mse = mean_squared_error(y_sim, y_pred)
    r2 = r2_score(y_sim, y_pred)
    
    results.append({
        'Lambda': lam,
        'Intercept': theta[0],
        'Slope': theta[1],
        'MSE': mse,
        'R2': r2})

pd.DataFrame(results)

,Lambda,Intercept,Slope,MSE,R2
0,0,1.137727,1.945275,1.949936,0.725824
1,1,1.137671,1.943850,1.949939,0.725823
2,10,1.137175,1.931119,1.950209,0.725785
3,100,1.132549,1.812414,1.974016,0.722438
4,1000,1.105658,1.122450,2.873516,0.595961
5,10000,1.071013,0.233509,5.947067,0.163796


λ = 0 (no regularization): Intercept ≈ 1, slope ≈ 2, matching the true parameters. This is ordinary least squares.
As λ increases: The slope shrinks toward 0 while the intercept stays near 1. MSE increases and R² decreases because the model is forced to use a smaller slope.
At very large λ (10000): The slope is nearly 0. The model predicts roughly the same value for every house, ignoring X entirely. R² drops close to 0.